### PR TITLE
Implemented bin chart and added it in compiler. 

### PR DESCRIPTION
**Pull request by anasdeis**
Implemented bin chart from scratch and added it in compiler. Added histogram/ in lib/ used for bin chart computations.

I looked into https://github.com/ytyeason/Visualization-on-Internet-of-Things for inspiration, but noticed many problems in their implementation such as:
 
They set x and y value for the axes (0 to 100) and hardcoded it. They use the bin size as a means to categorize bins, rather than setting the number of bins which can be done but is not typically the case for histograms. They determine the bin number by dividing the hardcoded axis range value by bin size which is simpler than setting the number of bins dynamically and in real-time by calculating the bin size based on the number of bins set manually or automatically. The maximum values for x and y change as they are being received which overcomplicates things as the old bin values are recategorized every time there is a change in the bin size due to a new maximum value received that changes the range of x and y axis. This issue has been dealt with. Therefore, I implemented code to compute the limits for the axes dynamically by setting the number of bins by the user manually or computing them automatically and in real-time using one of the methods set by the user ('squareRoot' | 'scott' | 'freedmanDiaconis' | 'sturges') with squareRoot being the default method. 

Here is a gif of what it looks like:

bin_chart_gif

### CODE REVIEW

**Comment by anasdeis** [comment]
This pull request resolves #29

**Comment by anasdeis** [comment]
This is basic bin chart implementation. This implementation is a basic version of count mode. As per the design proposal, there needs to be the customization options and their implementation. This implementation also does not include deviceID and the related history of device transitions which will be included in another issue.

**Comment by klaframboise** [comment]
Can you add the following on top of each file from the d3 array library. It is what the license demands https://github.com/d3/d3-array/blob/master/LICENSE

Copyright 2010-2020 Mike Bostock
All rights reserved.

**Comment by anasdeis** [comment]
I added the license to the @see references

**Comment by anasdeis** [comment]
I merged pull request 33 which implements the count mode fully as well as the history length

**Comment by klaframboise** [comment]
Changes broke transition period for bubble chart. (Sample 13)

```C:\Users\Kevin\dev\Treeflow.js\app\index.js:5
io.on('connection', socket => {console.log('connected');socket.on('message', body =>{    socket.broadcast.emit('message', {        body,        from: socket.id.slice(8)    });});socket.on('newDataPoint', body =>{    console.log('new data point received: '+body.x+','+body.y);    socket.broadcast.emit('newDataPoint', {        body,        from: socket.id.slice(8)    });});socket.on('newDataPoints', body =>{    console.log('new data array received, '+body.data.length+' new elements');    socket.broadcast.emit('newDataPoints', {        body,        from: socket.id.slice(8) 
   });});socket.on('disconnect', function() {    console.log('disconnect');});});





TypeError: Cannot read property 'length' of undefined
    at Socket.socket.on.body (C:\Users\Kevin\dev\Treeflow.js\app\index.js:5:463)
    at Socket.emit (events.js:189:13)
    at C:\Users\Kevin\dev\Treeflow.js\app\node_modules\socket.io\lib\socket.js:528:12
    at process._tickCallback (internal/process/next_tick.js:61:11)```

File: utils/code-blocks/binHistoryLengthInterval.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/*
+ * History length handling code. Every 30 seconds, check the list of device
+ * and remove any device that did not receive an update in the last historyLength.
+ */
+
+const historyLength = 120000;
```
Comment (klaframboise) [inline utils/code-blocks/binHistoryLengthInterval.js:6]: Looks hard coded. I don't see anywhere it being replaced by the config value, could be wrong though.

File: doc/Samples/13.Bubble-Chart-transition-period/Data-mocker.js [hunk 2]
```diff
@@ -20,7 +20,7 @@ const deviceDict = {
 };
 
 console.log('emitting initial dict');
-socket.emit('newDataPoints', {id: 0, centroids: Object.values(deviceDict)});
+socket.emit('newDataPoints', {id: 0, data: Object.values(deviceDict)});
```
Comment (klaframboise) [inline doc/Samples/13.Bubble-Chart-transition-period/Data-mocker.js:23]: Need to do the same for the emit on line 28

File: utils/EchartAdaptor.js [hunk 3]
```diff
@@ -297,6 +297,72 @@ module.exports = {
...
+    },
+
+    bin: function () {
+
+        const data = this.array.toJSON();
+        const data_heatmap = (data.length > 0)
+            ? computeBins2dHeatmap(data, null, 10, 10)
+            : [];
+
+        /*console.log('data')
```
Comment (klaframboise) [inline utils/EchartAdaptor.js:310]: Remove

File: utils/code-blocks/binHistoryLengthInterval.js (hunk 1 above)
Comment (anasdeis) [reply utils/code-blocks/binHistoryLengthInterval.js:6]: It is hardcoded, I implement the code for customization option in issue #35

File: doc/Samples/13.Bubble-Chart-transition-period/Data-mocker.js (hunk 2 above)
Comment (anasdeis) [reply doc/Samples/13.Bubble-Chart-transition-period/Data-mocker.js:23]: Fixed it with commit c3c0d5d49c3493b73a6be785733addc5e58540c8

File: utils/EchartAdaptor.js (hunk 3 above)
Comment (anasdeis) [reply utils/EchartAdaptor.js]: Removed it with commit c3c0d5d49c3493b73a6be785733addc5e58540c8

**Review by klaframboise — APPROVED** [review]
Thanks for the revisions!

**Pull request merged**